backstack/src/commonMain/kotlin/com/slack/circuit/backstack/BackStack.kt

```diff
@@ -53,6 +53,24 @@ public interface BackStack<R : Record> : Iterable<R> {
     while (topRecord?.let(predicate) == false) pop()
   }
 
+  /**
+   * Saves the current back stack entry list in an internal state store. It can be later restored by
+   * the root screen to [restoreState].
+   *
+   * This call will overwrite any existing stored state with the same root screen.
+   */
+  public fun saveState()
+
+  /**
+   * Restores the saved state with the given [screen], adding it on top of the existing entry list.
+   * If you wish to replace the current entry list, you should [pop] all of the existing entries off
+   * before calling this function.
+   *
+   * @param screen The root screen which was previously saved using [saveState].
+   * @return Returns true if there was any back stack state to restore.
+   */
+  public fun restoreState(screen: Screen): Boolean
```

Comment:
I wonder if we should use the record key instead? Implementation-wise, the screen is usually unique but only the key is guaranteed/required to be.

Reply:
We explicitly don't want uniqueness here though, we want equality. Otherwise we'll never be able to restore as we won't know the key of the previous instance of the screen.

Reply:
I did think about exposing the `key` as a parameter, but it gets messy with `SaveableBackStack` as the key needs to be `Any : Saveable`, which means adding another type parameter to `BackStack`

Reply:
Fair. Sidenote re: saveable - I'm not following, the key is a String no?

Reply:
Sorry I meant the `key` to the stateStore map, rather than hardcoding the `backstack.last().screen`